swift/ql/src/queries/Security/CWE-020/MissingRegexAnchor.qhelp

```diff
@@ -0,0 +1,74 @@
+<!DOCTYPE qhelp PUBLIC
+"-//Semmle//qhelp//EN"
+"qhelp.dtd">
+<qhelp>
+
+	<overview>
+		<p>
+
+			Sanitizing untrusted input with regular expressions is a
+			common technique.  However, it is error-prone to match untrusted input
+			against regular expressions without anchors such as <code>^</code> or
+			<code>$</code>.  Malicious input can bypass such security checks by
+			embedding one of the allowed patterns in an unexpected location.
+
+		</p>
+
+		<p>
+
+			Even if the matching is not done in a security-critical
+			context, it may still cause undesirable behavior when the regular
+			expression accidentally matches.
+
+		</p>
+	</overview>
+
+	<recommendation>
+		<p>
+
+			Use anchors to ensure that regular expressions match at
+			the expected locations.
+
+		</p>
+	</recommendation>
+
+	<example>
+
+		<p>
+
+			The following example code checks that a URL redirection
+			will reach the <code>example.com</code> domain, and not
+			some malicious site.
+
+		</p>
+
+		<sample src="MissingRegexAnchorBad.swift"/>
+
+		<p>
+
+			The check with the regular expression match is, however, easy to bypass. For example
+			by embedding <code>http://www.example.com/</code> in the query
+			string component: <code>http://evil-example.net/?x=http://www.example.com/</code>.
+			Address these shortcomings by using anchors in the regular expression instead:
+
+		</p>
+
+		<sample src="MissingRegexAnchorGood.swift"/>
+
+		<p>
+
+			A related mistake is to write a regular expression with
+			multiple alternatives, but to only include an anchor for one of the
+			alternatives. As an example, the regular expression
+			<code>/^www\.example\.com|beta\.example\.com/</code> will match the host
+			<code>evil.beta.example.com</code> because the regular expression is parsed
+			as <code>/(^www\.example\.com)|(beta\.example\.com)/</code>
```

Comment:
```suggestion
			If you need to write a regular expression to match
			multiple hosts, you should include an anchor for all of the
			alternatives. For example, the regular expression
			<code>/^www\.example\.com|beta\.example\.com/</code> will only match the host
			<code>evil.beta.example.com</code>, because the regular expression is parsed
			as <code>/(^www\.example\.com)|(beta\.example\.com)/</code>.
```